Adding an e2e test on Windows GMSA support

wk8 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This patch is adding an e2e on the Windows GMSA support added in
https://github.com/kubernetes/kubernetes/pull/73726 and
https://github.com/kubernetes/kubernetes/pull/74737.

**Which issue(s) this PR fixes**:
@yujuhong requested an e2e test on this at https://github.com/kubernetes/kubernetes/pull/73726#pullrequestreview-206595783.

**Special notes for your reviewer**:
This e2e test has been verified to pass on a cluster with a Windows node with the
`WindowsGMSA=true` feature gate enabled (and the kubelet compiled from
https://github.com/kubernetes/kubernetes/pull/74737); however, I don't believe that
feature gate is enabled when running a "regular" build. Would it be possible to:
1) enable that feature gate when running Windows e2e tests in a build?
2) or else enable dynamic Kubelet config for these tests, so that each spec can
set the config it needs?

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

wk8 commented:
@yujuhong @dchen1107 @michmike @PatrickLang : guidance appreciated as to how to make sure that e2e test can run successfully as part of builds (see the `Special notes for your reviewer` section in the  PR message). Thanks!

wk8 commented on test/e2e/windows/density.go:
```diff
@@ -30,21 +30,14 @@ import (
-	BeforeEach(func() {
-		// NOTE(vyta): these tests are Windows specific
-		framework.SkipUnlessNodeOSDistroIs("windows")
-	})
```
Note to reviewers: this has been moved to https://github.com/kubernetes/kubernetes/pull/74738/files#diff-b0ddb1c276f901c22122f6283cc7348c

ddebroy commented:
/ok-to-test

ddebroy commented:
/assign dchen1107

ddebroy commented:
/assign @adelina-t @bclau

ddebroy commented:
/sig windows

neolit123 reviewed (commented):
/area test
/kind feature
/priority important-soon

michmike commented:
@adelina-t  and @bclau  can you please read the comments above regarding how to light up this test in automation

michmike commented:
@neolit123 can you also please review the test?

ddebroy commented on test/e2e/windows/gmsa.go:
```diff
@@ -0,0 +1,160 @@
+// I'm afraid this is the smallest one that does.
+const gMSATestImageName = "mcr.microsoft.com/powershell"
+
+var _ = SIGDescribe("[Feature:Windows] GMSA [Slow]", func() {
```
The Feature tag above, I think, should also refer to the GMSA feature tag.

wk8 replied:
👍

neolit123 commented on test/e2e/windows/gmsa.go:
```diff
@@ -0,0 +1,160 @@
+
+// We need an image that packs the `nltest` command for this - and the `winPauseImageName` one doesn't.
+// I'm afraid this is the smallest one that does.
+const gMSATestImageName = "mcr.microsoft.com/powershell"
```
yeah too bad nltest is not part of the nanoserver images...
but also nl*.dll from system32 isn't either.

adelina-t replied:
Same comment as above. No need to hardcore. We should just use one of the images in "k8s.io/kubernetes/test/utils/image", they are all based on windowsservercore , thus will have powershell. If not, we can just define another image in "k8s.io/kubernetes/test/utils/image", either way I suggest we stick to having all image names and repos there.

wk8 replied:
Neat, hadn't seen the trick of overriding the repos. Thanks.

neolit123 commented on test/e2e/windows/gmsa.go:
```diff
@@ -0,0 +1,160 @@
+
+	Describe("kubelet GMSA support", func() {
+		Context("when creating a pod with correct GMSA credential specs", func() {
+			It("passes the credential specs down to the pod's containers", func() {
```
non blocking nit: ideally `pod`-> `Pod` in all user facing strings.

wk8 replied:
👍

neolit123 reviewed (commented):
LGTM
as much as i understand about nltest.exe and the AD config that is tested here.

michmike commented:
/lgtm

michmike commented:
/assign @dchen1107

adelina-t commented on test/e2e/windows/framework.go:
```diff
@@ -16,8 +16,23 @@ limitations under the License.
+
+// it seems we can't use k8s.io/kubernetes/test/utils/image.GetPauseImageName() as it does not
+// contain a windows image?
+const winPauseImageName = "kubeletwin/pause:latest"
```
Not a fan of this being hardcoded here.

Also, the pause image here ( the one returned by GetPauseImageName() ): https://github.com/kubernetes/kubernetes/blob/b67126e350fdba6d89a0ee64310cac7db14e33d3/test/utils/image/manifest.go#L128 can be a windows image. 

By passing this env var: https://github.com/kubernetes/kubernetes/blob/b67126e350fdba6d89a0ee64310cac7db14e33d3/test/utils/image/manifest.go#L66 

with a file containing the registry names for windows images, the pause image should be fine.

In the upstream tests we use these registries: https://github.com/kubernetes-sigs/windows-testing/blob/master/images/image-repo-list

In fact, all the dockerfiles + documentation of the images we use in upstream e2e windows tests is here:

https://github.com/kubernetes-sigs/windows-testing

wk8 replied:
👍

adelina-t commented on test/e2e/windows/density.go:
```diff
@@ -30,21 +30,14 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/kubernetes/test/e2e/framework"
-	imageutils "k8s.io/kubernetes/test/utils/image"
```
All image names should be imported from here for consistency

wk8 replied:
👍

adelina-t reviewed (commented):
Thank you for this PR :). Left some comments inside.

wk8 commented:
Thanks for the review @adelina-t  :) addressed your comments.

bclau commented on test/e2e/windows/README.md:
````diff
@@ -4,7 +4,7 @@
 ```bash
 KUBECONFIG=path/to/kubeconfig
-curl https://raw.githubusercontent.com/e2e-win/e2e-win-prow-deployment/master/repo-list -o repo_list
+curl https://raw.githubusercontent.com/e2e-win/e2e-win-prow-deployment/master/repo-list -o repo_list # or https://raw.githubusercontent.com/kubernetes-sigs/windows-testing/master/images/image-repo-list
````
Considering that kubernetes-sigs/windows-testing is the official repo, that would make it's image-repo-list file official and up to date. I would just have that file mentioned, no need for both.

wk8 replied:
👍

bclau commented on test/e2e/windows/gmsa.go:
```diff
@@ -0,0 +1,160 @@
+						Name:    name,
+						Image:   imageutils.GetPauseImageName(),
+						Command: []string{"powershell"},
+						Args:    []string{"while ($true) { Start-Sleep -Seconds 1 }"},
```
Well, the command and args are unnecessary. It is a pause image after all, it is supposed to do that anyways.

wk8 replied:
Good catch, thanks. Changed.

Sorry I forgot changing this as part of 9eac1bebd024ccf510ab126c08d7dfacd1e651d0 (was using image `mcr.microsoft.com/powershell` before).

ddebroy commented:
/test pull-kubernetes-e2e-gc

ddebroy commented:
/test pull-kubernetes-e2e-gce

ddebroy commented on test/e2e/windows/README.md:
````diff
@@ -4,7 +4,7 @@
 ```bash
 KUBECONFIG=path/to/kubeconfig
-curl https://raw.githubusercontent.com/e2e-win/e2e-win-prow-deployment/master/repo-list -o repo_list
+curl https://raw.githubusercontent.com/kubernetes-sigs/windows-testing/master/images/image-repo-list -o repo_list
````
Can this be moved to a separate commit from the main GMSA test commit?

wk8 replied:
👍

ddebroy commented on test/e2e/windows/gmsa.go:
```diff
@@ -0,0 +1,158 @@
+				// have been properly cleaned up - but as of right now, e2e_node tests don't support Windows. We should migrate this
+				// test to an e2e_node test when they start supporting Windows.
+
+				// FIXME: remove this before merging - this is here to check if this test is actually run
```
This needs to be removed/fixed right?

wk8 replied:
Yes - but before I remove it I'd like to see it failing where it's supposed to be running, to make sure this test actually runs somewhere... And I haven't seen that yet?

ddebroy replied:
Discussed with @wk8 offline that these tests won't run as PROW or get picked up by the other testgrid suites right now.

ddebroy reviewed (commented):
Looks good overall. Please address the FIX comment, squash the commits and separate out the README fix to a separate commit.

wk8 committed:
Adding an e2e test on Windows GMSA support

This patch is adding an e2e on the Windows GMSA support added in
https://github.com/kubernetes/kubernetes/pull/73726 and
https://github.com/kubernetes/kubernetes/pull/74737.

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 committed:
Updating Windows' e2e tests' README to the official repo list

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

ddebroy commented:
/lgtm

ddebroy commented:
/test pull-kubernetes-e2e-gce-100-performance

michmike commented:
/approve

michmike commented:
/lgtm

yujuhong commented on test/e2e/windows/gmsa.go:
```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
s/2018/2019

yujuhong commented on test/e2e/windows/gmsa.go:
```diff
@@ -0,0 +1,155 @@
+				f.PodClient().Create(pod)
+
+				ginkgo.By("waiting for the pod and its containers to be running")
+				gomega.Eventually(func() bool {
```
You can just use `framework.WaitForPodReady()`
There are other utilities functions  like that. Is there any reason to roll your own?

wk8 replied:
No reason other than I didn't spot these functions, apologies. @adelina-t kindly fixed that in https://github.com/kubernetes/kubernetes/pull/75116
